File: playground-swift/Sources/DataStructure/ListNode.swift
public class ListNode {
    public var val: Int
    public var next: ListNode?
    public init() { self.val = 0; self.next = nil; }
    public init(_ val: Int) { self.val = val; self.next = nil; }
    public init(_ val: Int, _ next: ListNode?) { self.val = val; self.next = next; }

    public static func fromArray(_ array: [Int]) -> ListNode? {
        guard !array.isEmpty else { return nil }
        let head = ListNode(array[0])
        var curr = head
        for i in 1..<array.count {
            curr.next = ListNode(array[i])
            curr = curr.next!
        }
        return head
    }

    public func toArray() -> [Int] {
        var res = [Int]()
        var curr: ListNode? = self
        while curr != nil {
            res.append(curr!.val)
            curr = curr!.next
        }
        return res
    }
}

File: playground-cpp/src/datastructure/list_node.cpp
#include "datastructure/list_node.hpp"

ListNode* ListNode::fromVector(const std::vector<int>& nums) {
    if (nums.empty()) return nullptr;
    ListNode* head = new ListNode(nums[0]);
    ListNode* curr = head;
    for (size_t i = 1; i < nums.size(); ++i) {
        curr->next = new ListNode(nums[i]);
        curr = curr->next;
    }
    return head;
}

std::vector<int> ListNode::toVector(ListNode* head) {
    std::vector<int> res;
    while (head) {
        res.push_back(head->val);
        head = head->next;
    }
    return res;
}

void ListNode::freeList(ListNode* head) {
    while (head) {
        ListNode* next = head->next;
        delete head;
        head = next;
    }
}
